File: src/CodeLearn.Domain/Exercises/MethodCodingExercise.cs
namespace CodeLearn.Domain.Exercises;

public sealed class MethodCodingExercise : Exercise
{
    public string MethodToExecute { get; private set; }
    public string MethodSolutionCode { get; private set; }
    public DataTypeId MethodReturnDataTypeId { get; private set; } = null!;
    public DataType MethodReturnDataType { get; private set; } = null!;

    private readonly IList<MethodParameter> _methodParameters = [];
    public IReadOnlyList<MethodParameter> MethodParameters => _methodParameters.AsReadOnly();

    private readonly IList<TestCase> _testCases = [];
    public IReadOnlyList<TestCase> TestCases => _testCases.AsReadOnly();

    private readonly IList<InputOutputExample> _inputOutputExamples = [];
    public IReadOnlyList<InputOutputExample> InputOutputExamples => _inputOutputExamples.AsReadOnly();

    private MethodCodingExercise(
        TestId testId,
        string title,
        string description,
        ExerciseDifficulty difficulty,
        string methodToExecute,
        string methodSolutionCode)
        : base(testId, title, description, difficulty)
    {
        MethodToExecute = methodToExecute;
        MethodSolutionCode = methodSolutionCode;
    }

    public static MethodCodingExercise Create(
        TestId testId,
        string title,
        string description,
        ExerciseDifficulty difficulty,
        string methodToExecute,
        string methodSolutionCode,
        DataType dataType)
    {
        var exercise = new MethodCodingExercise(
           testId,
           title,
           description,
           difficulty,
           methodToExecute,
           methodSolutionCode)
        {
            MethodReturnDataType = dataType,
            MethodReturnDataTypeId = dataType.Id
        };

        return exercise;
    }

    public void AddMethodParameter(MethodParameter methodParameter)
    {
        _methodParameters.Add(methodParameter);
    }

    public void AddTestCase(TestCase testCase)
    {
        _testCases.Add(testCase);
    }

    public void AddExample(InputOutputExample example)
    {
        _inputOutputExamples.Add(example);
    }
}

File: src/CodeLearn.WebApp/src/features/dashboard/tests/models/MethodCodingExercise.ts
import { MethodParameter } from '@/features/dashboard/tests/models/MethodParameter.ts';
import { TestCase } from '@/features/dashboard/tests/models/TestCase.ts';
import { InputOutputExample } from '@/features/dashboard/tests/models/InputOutputExample.ts';
import { ExerciseDifficulty } from '@/features/dashboard/tests/models/ExerciseDifficulty.ts';
import { ExerciseNote } from '@/features/dashboard/tests/models/ExerciseNote.ts';
import { ExerciseTopic } from '@/features/dashboard/tests/models/ExerciseTopic.ts';

export type MethodCodingExercise = {
  // Exercise fields
  id: number;
  testId: number;
  title: string;
  description: string;
  difficulty: ExerciseDifficulty;
  exerciseTopics?: ExerciseTopic[];
  // Method coding fields
  methodToExecute: string;
  methodSolutionCode?: string;
  methodReturnDataTypeId: number;
  exerciseNotes?: ExerciseNote[];
  methodParameters: MethodParameter[];
  testCases: TestCase[];
  inputOutputExamples?: InputOutputExample[];
};
